### DISCUSSION TITLE
Source Generator to Create Concrete Constant<T> Subclasses

### DISCUSSION THREAD

**Discussion by zacharylayne** [dotnet/dotNext#191] [Ideas]
Really loving the constant generic arguments, but it feels like busy work to define `Constant<T>` subclasses. I'd like to suggest creating a source generator to do that instead.

My idea is that usage would be similar to the following code:

````
[Constant<ulong>(UInt64.Max, "Max")]
[Constant<ulong>(0, "Zero")]
public partial abstract class UInt64Const : Constant<ulong>() { }
````

The `ConstantAttribute<T>` must be applied to a partial abstract class that derives from `Constant<T>`. This would create nested subclasses with their names & constant values specified along with the attribute. The attribute would have a static constructor to validate that `T` values can be constant, and a protected instance constructor to validate class names.

Feel free to let me know if this is a dumb idea! :|

**Comment by sakno** [comment]
Currently, I'm not investing time in this feature. .NET 8 is coming, and starting from November I'll be focused on releasing next major version of .NEXT 5.x. Since .NET 7 (and even .NET 6, with enabled preview features) constant generic argument can be efficiently replaced with static abstract interface property, e.g.:
```csharp
public interface IConstant<out T>
{
    static abstract T Value { get; }
}

static void MethodWithGenericConstant<T>() where T : IConstant<int>
{
    int value = T.Value;
}
```

Therefore, I have a plan to remove constant generic arguments in the next version.

**Reply by sakno** [reply]
Related: dotnet/runtime#89730
